src/main/java/com/amazon/jenkins/ec2fleet/EC2FleetCloud.java

```diff
@@ -1043,4 +1035,115 @@ public boolean configure(final StaplerRequest req, final JSONObject formData) th
+        public static class DescriptorImpl extends ExecutorScaleDescriptor {
+            @Override
+            public String getDisplayName() { return "No scaling"; }
+        }
+    }
+
+    public static class WeightedScaler extends ExecutorScaler {
+
+        @DataBoundConstructor
+        public WeightedScaler() {}
+        @Override
+        public int scale(String instanceType, FleetStateStats stats, AmazonEC2 ec2) {
+            if(stats == null) {
+                return numExecutors;
+            }
+
+            final Double instanceTypeWeight = stats.getInstanceTypeWeights().get(instanceType);
+            if (instanceTypeWeight == null) {
+                return numExecutors;
+            }
+            return (int) Math.max(Math.round(numExecutors * instanceTypeWeight), 1);
+        }
+
+        @Extension
+        public static class DescriptorImpl extends ExecutorScaleDescriptor {
+            @Override
+            public String getDisplayName() { return "Scale by weight";}
+        }
+    }
+
+    public static class NodeHardwareScaler extends ExecutorScaler {
+        private int vCpuPerExecutor;
+        private int memoryGiBPerExecutor;
+
+        @DataBoundConstructor
+        public NodeHardwareScaler(int vCpuPerExecutor, int memoryGiBPerExecutor) {
+            this.vCpuPerExecutor = vCpuPerExecutor;
+            this.memoryGiBPerExecutor = memoryGiBPerExecutor;
+        }
+
+        @DataBoundSetter
+        public void setvCpuPerExecutor(int value) { this.vCpuPerExecutor = value; }
+
+        @DataBoundSetter
+        public void setMemoryGiBPerExecutor(int value) { this.memoryGiBPerExecutor = value; }
+
+        @Override
+        public int scale(String instanceType, FleetStateStats stats, AmazonEC2 ec2) {
+            if(this.vCpuPerExecutor == 0 && this.memoryGiBPerExecutor == 0) {
+                return numExecutors;
+            }
+
+            int vCPUNumExecutors = Integer.MAX_VALUE;
+            int memoryNumExecutors = Integer.MAX_VALUE;
+            InstanceTypeInfo instanceTypeInfo = getInstanceTypeInfo(ec2, instanceType);
+            if(this.vCpuPerExecutor != 0) {
+                int instanceVCPUs = instanceTypeInfo.getVCpuInfo().getDefaultVCpus();
+                vCPUNumExecutors = Math.max(instanceVCPUs/this.vCpuPerExecutor, 1);
+            }
+
+            if(this.memoryGiBPerExecutor != 0) {
+                long instanceMemory = instanceTypeInfo.getMemoryInfo().getSizeInMiB()/1024;
+                memoryNumExecutors = Math.max(Math.round((float) instanceMemory /this.memoryGiBPerExecutor), 1);
+            }
+            return Math.min(vCPUNumExecutors, memoryNumExecutors);
+        }
+
+        @Extension
+        public static class DescriptorImpl extends ExecutorScaleDescriptor {
+            @Override
+            public String getDisplayName() { return "Scale by node hardware";}
+        }
+
+        private InstanceTypeInfo getInstanceTypeInfo(AmazonEC2 ec2, String instanceType) {
```

Comment:
For a fleet with large number of instance types, this will introduce one API call per instance type which is unnecessary. We should instead track the instanceTypeInfo in `FleetStateStats` just like how weights are populated [here](https://github.com/jenkinsci/ec2-fleet-plugin/blob/master/src/main/java/com/amazon/jenkins/ec2fleet/fleet/AutoScalingGroupFleet.java#L94-L107).

Reply:
On the other hand, this is the only place that works for InstanceRequirements use cases. Let's leave it here and come back to it as needed. :+1: